arbnode/transaction_streamer.go

```diff
@@ -218,68 +195,100 @@ func deleteStartingAt(db ethdb.Database, batch ethdb.Batch, prefix []byte, minKe
 
 // The insertion mutex must be held. This acquires the reorg mutex.
 // Note: oldMessages will be empty if reorgHook is nil
-func (s *TransactionStreamer) reorgToInternal(batch ethdb.Batch, count arbutil.MessageIndex) ([]*arbstate.MessageWithMetadata, error) {
+func (s *TransactionStreamer) reorg(batch ethdb.Batch, count arbutil.MessageIndex, newMessages []arbostypes.MessageWithMetadata) error {
 	if count == 0 {
-		return nil, errors.New("cannot reorg out init message")
+		return errors.New("cannot reorg out init message")
+	}
+	lastDelayedSeqNum, err := s.getPrevPrevDelayedRead(count)
+	if err != nil {
+		return err
+	}
+	var oldMessages []*arbostypes.MessageWithMetadata
+
+	targetMsgCount, err := s.GetMessageCount()
+	if err != nil {
+		return err
+	}
+	config := s.config()
+	maxResequenceMsgCount := count + arbutil.MessageIndex(config.MaxReorgResequenceDepth)
+	if config.MaxReorgResequenceDepth >= 0 && maxResequenceMsgCount < targetMsgCount {
+		log.Error(
+			"unable to re-sequence all old messages because there are too many",
+			"reorgingToCount", count,
+			"removingMessages", targetMsgCount-count,
+			"maxReorgResequenceDepth", config.MaxReorgResequenceDepth,
+		)
+		targetMsgCount = maxResequenceMsgCount
 	}
-	var oldMessages []*arbstate.MessageWithMetadata
-	if s.reorgSequencing {
-		targetMsgCount, err := s.GetMessageCount()
+	for i := count; i < targetMsgCount; i++ {
+		oldMessage, err := s.GetMessage(i)
 		if err != nil {
-			return nil, err
+			log.Error("unable to lookup old message for re-sequencing", "position", i, "err", err)
+			break
 		}
-		config := s.config()
-		maxResequenceMsgCount := count + arbutil.MessageIndex(config.MaxReorgResequenceDepth)
-		if config.MaxReorgResequenceDepth >= 0 && maxResequenceMsgCount < targetMsgCount {
-			log.Error(
-				"unable to re-sequence all old messages because there are too many",
-				"reorgingToCount", count,
-				"removingMessages", targetMsgCount-count,
-				"maxReorgResequenceDepth", config.MaxReorgResequenceDepth,
-			)
-			targetMsgCount = maxResequenceMsgCount
+
+		if oldMessage.Message == nil || oldMessage.Message.Header == nil {
+			continue
 		}
-		for i := count; i < targetMsgCount; i++ {
-			oldMessage, err := s.GetMessage(i)
-			if err != nil {
-				log.Error("unable to lookup old message for re-sequencing", "position", i, "err", err)
-				break
+
+		header := oldMessage.Message.Header
+
+		if header.RequestId != nil {
+			// This is a delayed message
+			delayedSeqNum := header.RequestId.Big().Uint64()
+			if delayedSeqNum+1 != oldMessage.DelayedMessagesRead {
+				log.Error("delayed message header RequestId doesn't match database DelayedMessagesRead", "header", oldMessage.Message.Header, "delayedMessagesRead", oldMessage.DelayedMessagesRead)
+				continue
+			}
+			if delayedSeqNum != lastDelayedSeqNum {
+				// This is the wrong position for the delayed message
+				continue
+			}
+			lastDelayedSeqNum++
```

Comment:
I think we want to increment `lastDelayedSeqNum++` after the `if s.inboxReader != nil` if statement just below. If we skipped resequencing delayed message 5 because it didn't match what the inbox tracker has, I don't think we then want to resequence delayed message 6 because that'd be out of order.